### PR TITLE
soc: telink_b91: Fix ROM region section overlap

### LINKED ISSUE #49036: soc: telink_b91: ROM region section overlap
**Describe the bug**

The Telink B91 SoC linker script defines its own `ROM_INIT` memory region that overlaps the `ROM` memory region defined by the [RISC-V arch common linker script](https://github.com/zephyrproject-rtos/zephyr/blob/5281b58d298066230b47b161d8926e3691fac22a/include/zephyr/arch/riscv/common/linker.ld#L79):

https://github.com/zephyrproject-rtos/zephyr/blob/5281b58d298066230b47b161d8926e3691fac22a/soc/riscv/riscv-privilege/telink_b91/linker.ld#L16

The intent of this seems to have been to place the `.init.*` sections at the beginning of the ROM memory region, and it achieves exactly that with the Binutils 2.35 included in the Zephyr SDK 0.14.2.

With the Binutils 2.38 included in the Zephyr SDK 0.15.0, however, the linker attempts to place these sections at the same memory address (because there are two separate overlapping memory regions) and outputs an error message saying that there are overlapping sections:

```
ld.bfd: section vector LMA [0000000020000000,0000000020000017] overlaps section vector LMA [0000000020000000,00000000200000bb]
```

**This is a bug, although it works for Binutils 2.35, because defining overlapping memory regions should not have been done in the first place and this working in the older versions of Binutils was purely by luck.**

The "correct" way of doing this would be to use the `zephyr_linker_sources(ROM_START ...`; but, the RISC-V linker script currently (questionably) places the `ROM_START` sections after the `vector` sections, which is not what this SoC wants to do:

https://github.com/zephyrproject-rtos/zephyr/blob/5281b58d298066230b47b161d8926e3691fac22a/include/zephyr/arch/riscv/common/linker.ld#L110-L126

**To Reproduce**

Build any test for `tlsr9518adk80d`.

```
scripts/twister -v -N -M -p tlsr9518adk80d -s tests/kernel/common/kernel.common
```

**Expected behavior**

Linker script does not define any overlapping memory regions.

**Impact**

Any boards based on the `telink_b91` SoC cannot be built with the Zephyr SDK 0.15.0 and above.

**Logs and console output**

```
/opt/zephyr-sdk-0.15.0-rc1/riscv64-zephyr-elf/bin/../lib/gcc/riscv64-zephyr-elf/12.1.0/../../../../riscv64-zephyr-elf/bin/ld.bfd: section vector LMA [0000000020000000,0000000020000017] overlaps section vector LMA [0000000020000000,00000000200000bb]
```

**Environment (please complete the following information):**

 - OS: Ubuntu 20.04
 - Toolchain: Zephyr SDK 0.15.0
 - Commit SHA: 2c301fe22f53468261246fb7cfd88b2092372d78

### PR DESCRIPTION
**Pull request by msierszulski**
This PR places the `vectors` sections after the `ROM_START` sections in [arch/riscv/common/linker.ld](https://github.com/zephyrproject-rtos/zephyr/blob/main/include/zephyr/arch/riscv/common/linker.ld)
It also adds an `init.ld` script that will prevent overlapping `.init` sections in the `telink_b91` SoC.

Fixes https://github.com/zephyrproject-rtos/zephyr/issues/49036.

### CODE REVIEW

File: soc/riscv/riscv-privilege/telink_b91/linker.ld [hunk 1]
```diff
@@ -17,16 +17,6 @@ MEMORY
...
 }
 
-SECTIONS
-{
-	SECTION_PROLOGUE(vector,,)
-	{
-		. = CONFIG_ROM_START_OFFSET;
-		. = ALIGN(4);
-		KEEP(*(.init.*))
-	} GROUP_LINK_IN(ROM_INIT)
```
Comment (stephanosio) [inline soc/riscv/riscv-privilege/telink_b91/linker.ld:27]: should remove `ROM_INIT (rx)  : ORIGIN = DT_REG_ADDR(DT_CHOSEN(zephyr_flash)), LENGTH = DT_REG_SIZE(DT_CHOSEN(zephyr_flash))` in L16 as well

Comment (msierszulski) [reply soc/riscv/riscv-privilege/telink_b91/linker.ld:27]: Thanks! I removed it.

**Review by stephanosio — DISMISSED** [review]

**Review by stephanosio — COMMENTED** [review]
just realised there is one more thing that needs to be updated:
https://github.com/zephyrproject-rtos/zephyr/blob/c029b081cc595415235045eaea9c923bcc4ed62e/cmake/modules/extensions.cmake#L1142-L1145

> Note: On RISC-V the rom_start section will be after vector section.

**Review by stephanosio — APPROVED** [review]

**Review by carlocaione — APPROVED** [review]

**Review by tgorochowik — APPROVED** [review]

**Comment by stephanosio** [comment]
It looks like there are some platforms and/or tests that actually rely on the `vectors` to come before `rom_start`.

https://github.com/zephyrproject-rtos/zephyr/runs/8158196183?check_suite_focus=true

* tests/kernel/gen_isr_table/arch.interrupt.gen_isr_table.riscv_direct (qemu_riscv32_smp:tests/kernel/gen_isr_table/arch.interrupt.gen_isr_table.riscv_direct) failed
* tests/kernel/gen_isr_table/arch.interrupt.gen_isr_table.riscv_direct (qemu_riscv32:tests/kernel/gen_isr_table/arch.interrupt.gen_isr_table.riscv_direct) failed
* tests/kernel/gen_isr_table/arch.interrupt.gen_isr_table.riscv_direct (qemu_riscv64_smp:tests/kernel/gen_isr_table/arch.interrupt.gen_isr_table.riscv_direct) failed
* tests/kernel/gen_isr_table/arch.interrupt.gen_isr_table.riscv_direct (qemu_riscv32e:tests/kernel/gen_isr_table/arch.interrupt.gen_isr_table.riscv_direct) failed
* tests/kernel/gen_isr_table/arch.interrupt.gen_isr_table.riscv_direct (qemu_riscv64:tests/kernel/gen_isr_table/arch.interrupt.gen_isr_table.riscv_direct) failed

**Comment by stephanosio** [comment]
> It looks like there are some platforms and/or tests that actually rely on the `vectors` to come before `rom_start`.

So the problem is that the failing tests set `CONFIG_GEN_IRQ_VECTOR_TABLE=y` and that places the IRQ vector table in the `ROM_START` section, which now precedes the `vectors` section.

https://github.com/zephyrproject-rtos/zephyr/blob/8563744997a057780a30b5566de8acfe699b3f9d/arch/common/CMakeLists.txt#L27-L31

Obviously, that is not working to work.

**Comment by fabiobaltieri** [comment]
> > It looks like there are some platforms and/or tests that actually rely on the `vectors` to come before `rom_start`.
> 
> So the problem is that the failing tests set `CONFIG_GEN_IRQ_VECTOR_TABLE=y` and that places the IRQ vector table in the `ROM_START` section, which now precedes the `vectors` section.
> 
> https://github.com/zephyrproject-rtos/zephyr/blob/8563744997a057780a30b5566de8acfe699b3f9d/arch/common/CMakeLists.txt#L27-L31
> 
> Obviously, that is not working to work.

Should this PR be reverted for now?

**Comment by stephanosio** [comment]
> Should this PR be reverted for now?

We should just disable `tests/kernel/gen_isr_table/arch.interrupt.gen_isr_table.riscv_direct` for now since that is the only affected test. I will file a bug for this and send a patch to do that.

In the meantime, I will send a patch to fix this (I have idea on how to fix this -- by making the vectors section use the same `ROM_START` mechanism, as does the ARM arch port.

**Pull request merged**